src/nvim/highlight.c

```diff
@@ -347,6 +347,7 @@ void update_window_hl(win_T *wp, bool invalid)
 {
   int ns_id = wp->w_ns_hl;
 
+  int old_nfloat = wp->w_floating && wp->w_hl_attr_normal > 0 ? wp->w_hl_attr_normal : -1;
```

Comment:
unfortunately this doesn't work as this might be the first time update_window_hl() is called for the window, so there is no cached value to use.

I poked around a little and think this condition is the right one:
```diff
diff --git a/src/nvim/highlight.c b/src/nvim/highlight.c
index 8729c74ce8..8436cd04de 100644
--- a/src/nvim/highlight.c
+++ b/src/nvim/highlight.c
@@ -370,12 +370,14 @@ void update_window_hl(win_T *wp, bool invalid)
 
   // determine window specific background set in 'winhighlight'
   bool float_win = wp->w_floating && !wp->w_config.external;
-  if (float_win && hl_def[HLF_NFLOAT] != 0) {
+  if (float_win && hl_def[HLF_NFLOAT] != 0 && ns_id > 0) {
     wp->w_hl_attr_normal = hl_def[HLF_NFLOAT];
   } else if (hl_def[HLF_COUNT] > 0) {
     wp->w_hl_attr_normal = hl_def[HLF_COUNT];
+  } else if (float_win) {
+    wp->w_hl_attr_normal = HL_ATTR(HLF_NFLOAT) >= 0 ? HL_ATTR(HLF_NFLOAT) : highlight_attr[HLF_NFLOAT];
   } else {
-    wp->w_hl_attr_normal = float_win ? HL_ATTR(HLF_NFLOAT) : 0;
+    wp->w_hl_attr_normal = 0;
   }
 
   if (wp->w_floating) {
```

Reply:
oh right i ignored the `highlight_attr`  thanks

Reply:
emm i give it a try the red background still missing

Reply:
ah, sorry, it should be `HL_ATTR(HLF_NFLOAT) > 0` and not `HL_ATTR(HLF_NFLOAT) >= 0`, updated above